prow/pod-utils/gcs/options.go

```diff
@@ -0,0 +1,197 @@
+// a parameter and will have the prefix prepended
+// to their destination in GCS, so the caller can
+// operate relative to the base of the GCS dir.
+func (o *Options) Run(extra map[string]UploadFunc) error {
+	var builder RepoPathBuilder
+	switch pathStrategyType(o.PathStrategy) {
+	case pathStrategyExplicit:
+		builder = NewExplicitRepoPathBuilder()
+	case pathStrategyLegacy:
+		builder = NewLegacyRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	case pathStrategySingle:
+		builder = NewSingleDefaultRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	}
+
+	spec, err := pjutil.ResolveSpecFromEnv()
+	if err != nil {
+		return fmt.Errorf("could not resolve job spec: %v", err)
+	}
+
+	var gcsPath string
+	jobBasePath := PathForSpec(spec, builder)
+	if o.SubDir == "" {
+		gcsPath = jobBasePath
+	} else {
+		gcsPath = path.Join(jobBasePath, o.SubDir)
+	}
+
+	uploadTargets := map[string]UploadFunc{}
+
+	// ensure that an alias exists for any
+	// job we're uploading artifacts for
+	if alias := AliasForSpec(spec); alias != "" {
+		uploadTargets[alias] = DataUpload(strings.NewReader(jobBasePath))
+	}
+
+	for _, item := range o.Items {
+		info, err := os.Stat(item)
+		if err != nil {
+			logrus.Warnf("Encountered error in resolving items to upload for %s: %v", item, err)
+			continue
+		}
+		if info.IsDir() {
+			gatherArtifacts(item, gcsPath, uploadTargets)
+		} else {
+			uploadTargets[path.Join(gcsPath, info.Name())] = FileUpload(item)
+		}
+	}
+
+	for destination, upload := range extra {
+		uploadTargets[path.Join(gcsPath, destination)] = upload
+	}
+
+	if !o.DryRun {
+		ctx := context.Background()
+		gcsClient, err := storage.NewClient(ctx, option.WithCredentialsFile(o.GceCredentialsFile))
+		if err != nil {
+			return fmt.Errorf("could not connect to GCS: %v", err)
+		}
+
+		if err := Upload(gcsClient.Bucket(o.GcsBucket), uploadTargets); err != nil {
+			return fmt.Errorf("failed to upload to GCS: %v", err)
+		}
+	} else {
+		for destination := range uploadTargets {
+			logrus.WithField("dest", destination).Info("Would upload")
+		}
+	}
+
+	return nil
+}
+
+func gatherArtifacts(artifactDir, gcsPath string, uploadTargets map[string]UploadFunc) {
+	logrus.Printf("Gathering artifacts from artifact directory: %s", artifactDir)
+	filepath.Walk(artifactDir, func(fspath string, info os.FileInfo, err error) error {
+		if info == nil || info.IsDir() {
+			return nil
+		}
+
+		// we know path will be below artifactDir, but we can't
+		// communicate that to the filepath module. We can ignore
+		// this error as we can be certain it won't occur and best-
+		// effort upload is OK in any case
+		if relPath, err := filepath.Rel(artifactDir, fspath); err == nil {
+			logrus.Printf("Found %s in artifact directory. Uploading as %s\n", fspath, relPath)
+			uploadTargets[path.Join(gcsPath, relPath)] = FileUpload(fspath)
+		} else {
+			logrus.Warnf("Encountered error in relative path calculation for %s under %s: %v", fspath, artifactDir, err)
```

Comment:
side note: have we thought about how we plan to debug this?

Reply:
This should be impossible, see above comment.

Reply:
Er, I meant to ask more generally about debugging this tool "in prod".